R/teal_slice.R

```diff
@@ -268,22 +268,60 @@ c.teal_slice <- function(...) {
   ans
 }
 
+#' @keywords internal
+#' @export
+as.list.teal_slice <- function(x) {
+  checkmate::assert_class(x, "teal_slice")
+
+  formals <- if (inherits(x, "teal_slice_expr")) {
+    formals(filter_expr)
+  } else {
+    formals(filter_var)
+  }
+
+  x <- if (shiny::isRunning()) {
+    shiny::reactiveValuesToList(x)
+  } else {
+    shiny::isolate(shiny::reactiveValuesToList(x))
+  }
+
+  formal_args <- setdiff(names(formals), '...')
+  extra_args <- setdiff(names(x), formal_args)
+
+  x[c(formal_args, extra_args)]
+}
+
 
-# format method for teal_slice
 #' @export
+#' @param show_all `logical(1)` should parameters set to NULL be returned
+#' @param center `logical(1)` should the output be centered and trimmed
 #' @rdname teal_slice
 #' @keywords internal
 #'
-format.teal_slice <- function(x, show_all = FALSE, ...) {
+format.teal_slice <- function(x, show_all = FALSE, center = TRUE, ...) {
```

Comment:
@chlebowa `store()` uses format with `center=FALSE` that's why we have `center` at all in `format`

Reply:
Ahh, I see. Well, let me ask this then: when we say that `print` and `store` should have the same output, meaning what we see in the console should be the same as what we see when we open a file in an editor, do we mean **identical**? Can we not make the printed version a little prettier?

If we allow for a cosmetic difference, saving and loading could be made simpler, as we could have:
```
print teal_slice:
`teal_slice` |> as.list |> toJSON |> center_json |> format |> cat

save teal_slice:
`teal_slice` |> as.list |> toJSON |> ends up in list with other slices

print teal_slices
`format.teal_slices` would:
1. `teal_slices` |> lapply(as.list) |> lapply(toJSON) |> lapply(center_json) |> lapply(format) -> slices
2. attributes(`teal_slices`) |> (drop class) |> toJSON |> (some adjustment) -> attributes
3. append attributes to slices
4. |> cat

save teal_slices:
1. `teal_slices` |> lapply(as.list) |> lapply(toJSON) -> slices
2. attributes(`teal_slices`) |> (drop class) |> toJSON -> attributes
3. append attributes to slices
4. |> write_json
```
Does this make sense?

Reply:
yes and this is what basically is happening but the code that you wrote in print is already in format, and the code that you have in store is also already in format, so you only call format with an extra parameter that parametrizes whether the output is centered or not and for the storage it is not and for the print it is

Reply:
so what we prepared so far is like below: one `format.teal_slice` function to rule them all + 2 parameters (`show_all` and `center`) so that code within other functions is not redundant

```{R}
print teal_slice:
  `teal_slice` |> format(center=TRUE) |> cat(console)

save teal_slice:
  `teal_slice` |> format(center=FALSE) |> cat(file)

format teal_slice:
  `teal_slice` |> as.list |> toJSON |> center(TRUE/FALSE)



format.teal_slices: 
  1. `teal_slices` |> lapply(format, center)
  2. attributes(`teal_slices`) |> (drop class) |> toJSON |> (some adjustment) -> attributes
  3. append attributes to slices
  
print.teal_slices:
  `teal_slices` |> format(center=TRUE) |> cat(console)

store.teal_slices:
  `teal_slices` |> format(center=TRUE) |> cat(file)
```

Reply:
and the print is not identical to store since print is centered, and that's the whole point of center parameter : P

Reply:
if we dont want the `center` option, then we don't need `store` function and we can just use print and pass a `file` argument that can be passed to `print`'s `cat` call